tomcat/tomcat-jta/src/main/java/org/jboss/narayana/tomcat/jta/TransactionalDataSourceFactory.java

```diff
@@ -0,0 +1,183 @@
+import org.apache.commons.dbcp2.managed.ManagedDataSource;
+import org.apache.commons.pool2.impl.GenericObjectPool;
+import org.apache.commons.pool2.impl.GenericObjectPoolConfig;
+
+import javax.naming.Context;
+import javax.naming.Name;
+import javax.naming.RefAddr;
+import javax.naming.Reference;
+import javax.naming.spi.ObjectFactory;
+import javax.sql.XADataSource;
+import javax.transaction.TransactionManager;
+import javax.transaction.xa.XAResource;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.lang.reflect.Type;
+import java.sql.SQLException;
+import java.util.Hashtable;
+import java.util.Properties;
+
+/**
+ * @author <a href="mailto:zfeng@redhat.com>Zheng Feng</a>
+ */
+public class TransactionalDataSourceFactory implements ObjectFactory {
+    private static final String PROP_USERNAME = "username";
+    private static final String PROP_PASSWORD = "password";
+    private static final String PROP_MAX_TOTAL = "maxTotal";
+    private static final String PROP_MAX_IDLE = "maxIdle";
+    private static final String PROP_MIN_IDLE = "minIdle";
+
+    private static final String[] ALL_PROPERTIES = {
+            PROP_USERNAME,
+            PROP_PASSWORD,
+            PROP_MAX_TOTAL,
+            PROP_MAX_IDLE,
+            PROP_MIN_IDLE
+    };
+
+    @Override
+    public Object getObjectInstance(Object obj, Name name, Context context, Hashtable<?, ?> environment) throws Exception {
+
+        if (obj == null || !(obj instanceof Reference)) {
+            return null;
+        }
+
+        final Reference ref = (Reference) obj;
+        if (!"javax.sql.DataSource".equals(ref.getClassName())) {
+            return null;
+        }
+
+        final Properties properties = new Properties();
+        for (final String propertyName : ALL_PROPERTIES) {
+            final RefAddr ra = ref.get(propertyName);
+            if (ra != null) {
+                properties.setProperty(propertyName, ra.getContent().toString());
+            }
+        }
+
+        TransactionManager transactionManager = (TransactionManager) getReferenceObject(ref, context, "transactionManager");
+        XADataSource xaDataSource = (XADataSource) getReferenceObject(ref, context, "xaDataSource");
+
+        XARecoveryModule xaRecoveryModule = getXARecoveryModule();
+        if (xaRecoveryModule != null) {
+            xaRecoveryModule.addXAResourceRecoveryHelper( new XAResourceRecoveryHelper() {
+                @Override
+                public boolean initialise(String p) throws Exception {
+                    return true;
+                }
+
+                @Override
+                public XAResource[] getXAResources() throws Exception {
+                    try {
+                        String user = properties.getProperty(PROP_USERNAME);
+                        String password = properties.getProperty(PROP_PASSWORD);
+
+                        if (user != null && password != null) {
```

Comment:
Do any of the databases allow a null password?

Reply:
No, I think most of the database does not allow the null password.